src/lib/src/x86_arrops.cpp

```diff
@@ -0,0 +1,133 @@
+#include "arch_config.h"
+
+#ifdef DISTOPIA_X86_SSE4_1
+
+#include <cstddef>
+#include <cstdint>
+#include <immintrin.h>
+
+#include "arrops.h"
+#include "compiler_hints.h"
+#include "x86_gintrin.h"
+#include "x86_swizzle.h"
+#include "ops.h"
+
+namespace {
+
+template <typename T, typename U>
+constexpr bool IsAligned(const U* addr) {
+  return reinterpret_cast<std::uintptr_t>(addr) % sizeof(T) == 0;
+}
+
+constexpr std::size_t kBigVectorThreshold = 256 * 1024;
```

Comment:
What exactly does this do?

Reply:
I’ve added comments that hopefully help 🙂 Basically, the processor has to pause for a bit to switch on the long vector (256-bit and 512-bit) registers. This means that long vector operations can actually be slower for small tasks. I (somewhat arbitrarily) define a task as big if it outputs at least 256 KiB of data.

The streaming thing is similar: there is a special kind of store called a ‘streaming store’, which is faster for big tasks but slower for small tasks.

Reply:
Okay awesome. Would it help if we could detect cache sizes with CMake and have it be some multiple thereof? I know detecting the cache size can be done.

Reply:
I agree that we should detect cache size, although I would lean towards doing it at runtime. I’m worried about making this PR too big; are you happy to address it in a separate one?

Reply:
Yeah ill raise an issue.